scipy/signal/_spectral_py.py

```diff
@@ -240,81 +244,106 @@ def lombscargle(
-    # store a and b so that phase can be calculated outside loop, if necessary
-    a = np.empty_like(freqs)
-    b = np.empty_like(freqs)
-
     # if requested, perform precenter
     if precenter:
         y -= y.mean()
 
-    # calculate the single sum that does not depend on the frequency
-    Y_sum = (weights * y).sum()
+    # transform arrays
+    # row vector
+    freqs = freqs[np.newaxis, :]
+    # column vectors
+    x = x[:, np.newaxis]
+    y = y[:, np.newaxis]
+    weights = weights[:, np.newaxis]  # type: ignore
 
-    # loop over the frequencies
-    for i in range(freqs.shape[0]):
-        coswt[:] = np.cos(freqs[i] * x)
-        sinwt[:] = np.sin(freqs[i] * x)
+    weights_y = weights * y
+    Y_sum = (weights_y).sum()
 
-        # calculate CC, SS, and CS first
-        # these variable names are of form XX_hat in the paper, but dropping the suffix
-        # to reuse variables later whether floating_mean is True or False
-        CC = (weights * coswt * coswt).sum()
-        SS = 1 - CC  # trig identity: S^2 = 1 - C^2
-        CS = (weights * coswt * sinwt).sum()
+    freqst = freqs * x
+    coswt = np.cos(freqst)
+    sinwt = np.sin(freqst)
 
-        # now, redefine trig arrays to always be multiplied by the weights
-        coswt[:] = weights * coswt
-        sinwt[:] = weights * sinwt
+    CC = 2.0 * np.dot(weights.T, 0.5 - sinwt**2.0)
+    SS = 2.0 * np.dot(weights.T, coswt * sinwt)
 
-        # these are also of the form XX_hat in the paper
-        YC = (y * coswt).sum()
-        YS = (y * sinwt).sum()
+    if floating_mean:
+        C = np.dot(weights.T, coswt)
+        S = np.dot(weights.T, sinwt)
+        CC -= C * C - S * S
+        SS -= 2.0 * C * S
 
-        if floating_mean:
-            # calculate best-fit offset for each frequency independently (default)
-            C_sum = coswt.sum()
-            S_sum = sinwt.sum()
-            YC -= Y_sum * C_sum
-            YS -= Y_sum * S_sum
-            CC -= C_sum * C_sum
-            SS -= S_sum * S_sum
-            CS -= C_sum * S_sum
-
-        # determinate of the system of linear equations
-        D = CC * SS - CS * CS
-
-        # where: y(w) = a*cos(w) + b*sin(w) + c
-        a[i] = (YC * SS - YS * CS) / D
-        b[i] = (YS * CC - YC * CS) / D
-        #  c = Y_sum - a * C_sum - b * S_sum  # offset is not useful to return
-
-        # store final value as power in (y units)^2
-        pgram[i] = 2.0 * (a[i] * YC + b[i] * YS)
-
-    if normalize == "normalize":
+    # calculate tau (phase offset to eliminate CS variable)
+    tau = 0.5 * np.arctan2(SS, CC)
+    freqst_tau = freqst - tau
+
+    # now, coswt and sinwt are offset by tau
+    coswt = np.cos(freqst_tau)
+    sinwt = np.sin(freqst_tau)
+
+    CC = np.dot(weights.T, coswt * coswt)
+
+    # by definition, CC can only be [0, 1]
+    # to prevent division by zero errors, limit to [0+eps, 1-eps]
+    eps = np.finfo(dtype=y.dtype).eps
+    CC[CC == 1.0] = 1.0 - eps
+    CC[CC == 0.0] = eps
```

Comment:
Another nitpick: It might be more a little appropriate to use `np.finfo(dtype=y.dtype).epsneg` here, since it represents the next possible float _smaller_ instead of _larger_ than 1.0

Reply:
Ah, good point. I didn't know about that.

Reply:
Done.